src/nwb_conversion_tools/tools/roiextractors/roiextractors.py

```diff
@@ -95,53 +86,52 @@ def add_two_photon_series(imaging, nwbfile, metadata, buffer_size=10, use_times=
 
     Adds two photon series from imaging object as TwoPhotonSeries to nwbfile object.
     """
-    metadata = dict_deep_update(get_default_ophys_metadata(), metadata)
-    metadata = safe_update(metadata, get_nwb_imaging_metadata(imaging))
-    # Tests if ElectricalSeries already exists in acquisition
-    nwb_es_names = [ac for ac in nwbfile.acquisition]
-    opts = metadata["Ophys"]["TwoPhotonSeries"][0]
-    if opts["name"] not in nwb_es_names:
-        # retrieve device
-        device = nwbfile.devices[list(nwbfile.devices.keys())[0]]
-        metadata["Ophys"]["ImagingPlane"][0]["optical_channel"] = [
-            OpticalChannel(**i) for i in metadata["Ophys"]["ImagingPlane"][0]["optical_channel"]
-        ]
-        metadata["Ophys"]["ImagingPlane"][0] = safe_update(metadata["Ophys"]["ImagingPlane"][0], {"device": device})
+    metadata = dict_deep_update(get_nwb_imaging_metadata(imaging), metadata)
 
-        imaging_plane = nwbfile.create_imaging_plane(**metadata["Ophys"]["ImagingPlane"][0])
+    # Tests if TwoPhotonSeries already exists in acquisition
+    two_photon_series_metadata = metadata["Ophys"]["TwoPhotonSeries"][0]
+    two_photon_series_name = two_photon_series_metadata["name"]
+    acquisition_modules = [module for module in nwbfile.acquisition]
 
+    # Only add if TwoPhotonSeries is not present before
+    if two_photon_series_name not in acquisition_modules:
+
+        # Add the data
         def data_generator(imaging):
             for i in range(imaging.get_num_frames()):
-                yield imaging.get_frames(frame_idxs=[i]).T
+                yield imaging.get_frames(frame_idxs=[i]).squeeze().T
 
         data = H5DataIO(
             DataChunkIterator(data_generator(imaging), buffer_size=buffer_size),
             compression=True,
         )
+        two_p_series_kwargs = two_photon_series_metadata
```

Comment:
Does this form a reference or a copy to the original metadata dictionary?

If the former, we should cop/ydeepcopy this so we can modify it without modifying the original structure.

Reply:
Good point. This did not occur to me before.

My first thought was thinking that the metadata should end as closely as possible to the final form as it is written in the `nwb_file` but then it seems way more important than functions should not have side effects. I can see this producing a hard to debug error in the future and if we want the former thing we should sort this out somewhere else.

Should be sorted in the last commit.